### PR TITLE
Quantity input

### PR DESCRIPTION
**Pull request by spireken**
The demo is called QuantityInputWidget. The two labels show when the change and submit signals are triggered, and they also take focus. The acceptable range is [-9999, 9999]. The disabled input on the right is to demonstrate the disabled style.

### CODE REVIEW

File: Applications/Spire/Include/Spire/Ui/IntegerInputWidget.hpp [hunk 1]
```diff
@@ -1,29 +1,36 @@
...
       //! Constructs a ValueInputWidget.
       /*
-        \param min_value The lowest value that can be entered.
-        \param max_value The highest value that can be entered.
+        \param value The initial value to display.
         \param parent The parent widget.
       */
-      IntegerInputWidget(int min_value, int max_value,
-        QWidget* parent = nullptr);
+      IntegerInputWidget(int value, QWidget* parent = nullptr);
```
Comment (andrei-sa) [inline Applications/Spire/Include/Spire/Ui/IntegerInputWidget.hpp:19]: `explicit`

File: Applications/Spire/Source/Ui/DecimalInputWidget.cpp [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+  set_stylesheet(false, false);
+  setContextMenuPolicy(Qt::NoContextMenu);
+  connect(this, QOverload<double>::of(&QDoubleSpinBox::valueChanged), this,
+    &DecimalInputWidget::on_value_changed);
+  connect(lineEdit(), &QLineEdit::textEdited, this,
+    &DecimalInputWidget::on_text_edited);
+  lineEdit()->installEventFilter(this);
+}
+
+QString DecimalInputWidget::textFromValue(double value) const {
```
Comment (andrei-sa) [inline Applications/Spire/Source/Ui/DecimalInputWidget.cpp:48]: This method does not use `this` and can be static.

File: Applications/Spire/Source/Ui/DecimalInputWidget.cpp [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+  if(value == 0.0) {
```
Comment (andrei-sa) [inline Applications/Spire/Source/Ui/DecimalInputWidget.cpp:49]: Use literals like `0.` unless requested otherwise.

File: Applications/Spire/Source/Ui/DecimalInputWidget.cpp [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+    return "0";
+  }
+  QString str = QString::number(value, 'f', decimals());
```
Comment (andrei-sa) [inline Applications/Spire/Source/Ui/DecimalInputWidget.cpp:52]: Use `auto`.

File: Applications/Spire/Source/Ui/DecimalInputWidget.cpp [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+      lineEdit()->setText(textFromValue(value()));
+      m_last_valid_value = value();
+      m_submit_signal(m_last_valid_value);
+      return;
+    case Qt::Key_Up:
+      add_step(1, event->modifiers());
+      return;
+    case Qt::Key_Down:
+      add_step(-1, event->modifiers());
+      return;
```
Comment (andrei-sa) [inline Applications/Spire/Source/Ui/DecimalInputWidget.cpp:123]: Just to confirm, should all these branches bypass the call `QDoubleSpinBox::keyPressEvent(event)` below?

File: Applications/Spire/Source/Ui/DecimalInputWidget.cpp (hunk 2 above)
Comment (spireken) [reply Applications/Spire/Source/Ui/DecimalInputWidget.cpp:48]: The textFromValue() method is overriding a base class method. I did change it to be protected to match the base class' declaration.

File: Applications/Spire/Source/Ui/DecimalInputWidget.cpp (hunk 5 above)
Comment (spireken) [reply Applications/Spire/Source/Ui/DecimalInputWidget.cpp:123]: Yes, the base class keyPressEvent() will have some kind of side effect (like highlighting all of the text) that I'm trying to avoid.

File: Applications/Spire/Source/Ui/DecimalInputWidget.cpp [hunk 6]
```diff
@@ -46,10 +46,10 @@ DecimalInputWidget::DecimalInputWidget(double value, QWidget* parent)
 }
 
 QString DecimalInputWidget::textFromValue(double value) const {
-  if(value == 0.0) {
+  if(value == 0) {
```
Comment (andrei-sa) [inline Applications/Spire/Source/Ui/DecimalInputWidget.cpp:49]: 0 is an integer literal. When comparing a value against a literal, always try to supply the same type of the literal as that of the value.
"0." (zero period) is a double literal and is a short form of "0.0".

Comment (spireken) [reply Applications/Spire/Source/Ui/DecimalInputWidget.cpp:49]: If the output is the same I'd prefer to keep the trailing zero to make it clearer.

Comment (andrei-sa) [reply Applications/Spire/Source/Ui/DecimalInputWidget.cpp:49]: Okay, restore the original double literal. If trailing zeros are prohibited (I do not know if they are), Kamal will probably let you know.

File: Applications/Spire/Source/Ui/IntegerInputWidget.cpp [hunk 7]
```diff
@@ -8,82 +8,31 @@
 using namespace boost::signals2;
 using namespace Spire;
 
-namespace {
-  auto BUTTON_X() {
-    static auto x = scale_width(12);
-    return x;
-  }
+IntegerInputWidget::IntegerInputWidget(int value, QWidget* parent)
```
Comment (kranar) [inline Applications/Spire/Source/Ui/IntegerInputWidget.cpp:11]: What do you think about using composition instead of inheritance for this?

Comment (spireken) [reply Applications/Spire/Source/Ui/IntegerInputWidget.cpp:11]: Wouldn't it have the same problem if it was using a DecimalInputWidget?

Comment (kranar) [reply Applications/Spire/Source/Ui/IntegerInputWidget.cpp:11]: Can you rephrase this question?

Comment (spireken) [reply Applications/Spire/Source/Ui/IntegerInputWidget.cpp:11]: I'm guessing based on you asking about using composition versus inheritance that there is a way to reuse a common numeric input for each of the numeric types. I'm not sure how I can modify the default QSpinBox implementations to create a generic numeric input that can be reused.

I can create a numeric input that uses Quantity as the base type that can be reused (assuming Quantity can express all our required numeric types), but that would require redoing some work regarding editing and fixing up the text in the widget.

Comment (kranar) [reply Applications/Spire/Source/Ui/IntegerInputWidget.cpp:11]: ```QDoubleSpinBox``` can be used for all the widgets using composition instead of inheritance.

Comment (kranar) [reply Applications/Spire/Source/Ui/IntegerInputWidget.cpp:11]: You can inherit from ```QDoubleSpinBox``` to implement the ```DecimalInputWidget``` as you have done, and then reuse it for the ```QuantityInputWidget``` and ```IntegerInputWidget```.

Comment (spireken) [reply Applications/Spire/Source/Ui/IntegerInputWidget.cpp:11]: I don't understand how that's different from inheritance in this case. The Quantity input for instance would still have to convert a double to a quantity (and vice-versa) if it was using the DecimalInputWidget.

Comment (kranar) [reply Applications/Spire/Source/Ui/IntegerInputWidget.cpp:11]: No, you can not operate on the ```double```, you would need to operate on the ```QString``` to parse it into a ```Quantity```.

Comment (kranar) [reply Applications/Spire/Source/Ui/IntegerInputWidget.cpp:11]: https://github.com/spiretrading/nexus/blob/3151c02be77001d09b15b2fcc5a30d8572328f12/Nexus/Include/Nexus/Definitions/Quantity.hpp#L37

Comment (kranar) [reply Applications/Spire/Source/Ui/IntegerInputWidget.cpp:11]: Similarly when you implement the ```MoneyInputWidget``` in the future, you will also parse it into a ```Money```.

Comment (spireken) [reply Applications/Spire/Source/Ui/IntegerInputWidget.cpp:11]: In order to call setMinimum(double), would I have to do something like:

`
auto qty = Quantity(1.234567)
auto str = item_delegate.displayText(qty);
setMinimum(str.toDouble());
`

Comment (spireken) [reply Applications/Spire/Source/Ui/IntegerInputWidget.cpp:11]: Actually, displayText just casts the Quantity to a double anyway.

File: Applications/Spire/Include/Spire/Ui/DecimalInputWidget.hpp [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+      //! Constructs a DecimalInputWidget.
+      /*!
+        \param value The initial value to display.
+        \param parent The parent widget.
+      */
+      explicit DecimalInputWidget(double value, QWidget* parent = nullptr);
+
+      //! Connects a slot to the value change signal.
+      boost::signals2::connection connect_change_signal(
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/DecimalInputWidget.hpp:23]: You may be able to get rid of these signals and stick to using ```QDoubleSpinBox```'s signals. All this widget does is style a ```QDoubleSpinBox``` correct?

File: Applications/Spire/Include/Spire/Ui/IntegerInputWidget.hpp [hunk 9]
```diff
@@ -1,29 +1,36 @@
 #ifndef SPIRE_INTEGER_INPUT_WIDGET_HPP
 #define SPIRE_INTEGER_INPUT_WIDGET_HPP
-#include <QSpinBox>
+#include "Spire/Ui/DecimalInputWidget.hpp"
 
 namespace Spire {
 
   //! Represents a widget for inputting whole numbers.
-  class IntegerInputWidget : public QSpinBox {
+  class IntegerInputWidget : public DecimalInputWidget {
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/IntegerInputWidget.hpp:8]: I still have hesitations that this inheritance is the right choice, but you can leave it for now.

File: Applications/Spire/Include/Spire/Ui/QuantityInputWidget.hpp [hunk 10]
```diff
@@ -1,48 +1,61 @@
...
-#include "Spire/Ui/TextInputWidget.hpp"
-#include "Spire/Ui/Ui.hpp"
+#include "Spire/Ui/DecimalInputWidget.hpp"
 
 namespace Spire {
 
-  //! Displays an input box that accepts Quantities.
-  class QuantityInputWidget : public TextInputWidget {
+  //! Represents a widget for displaying and modifying Quantities.
+  class QuantityInputWidget : public QWidget {
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/QuantityInputWidget.hpp:9]: Have you considered inheriting from ```QAbstractSpinBox```?

File: Applications/Spire/Source/Ui/QuantityInputWidget.cpp [hunk 11]
```diff
@@ -1,46 +1,56 @@
...
+    const ValueSignal::slot_type& slot) const {
+  return m_submit_signal.connect(slot);
 }
 
-void QuantityInputWidget::on_line_edit_committed() {
-  if(!text().isEmpty()) {
-    m_committed_signal(*Quantity::FromValue(text().toStdString()));
-  }
+void QuantityInputWidget::on_editing_finished() {
+  m_submit_signal(*Quantity::FromValue(m_input_widget->text().toStdString()));
```
Comment (kranar) [inline Applications/Spire/Source/Ui/QuantityInputWidget.cpp:51]: You need to validate anything that you parse.

File: Applications/Spire/Source/Ui/QuantityInputWidget.cpp [hunk 12]
```diff
@@ -1,46 +1,56 @@
...
+}
+
+void QuantityInputWidget::set_maximum(Nexus::Quantity maximum) {
+  m_input_widget->setMaximum(static_cast<double>(maximum));
 }
 
 void QuantityInputWidget::set_value(Quantity value) {
-  setText(m_item_delegate.displayText(
-    QVariant::fromValue(Truncate(value, 0)), m_locale));
+  m_input_widget->setValue(static_cast<double>(value));
```
Comment (kranar) [inline Applications/Spire/Source/Ui/QuantityInputWidget.cpp:37]: This isn't possible, not all ```Quantity```'s can be expressed as a ```double```. This is a big issue because even a common value like ```1.1``` is a valid ```Quantity``` but not a valid ```double```.

File: Applications/Spire/Include/Spire/Ui/DecimalInputWidget.hpp (hunk 8 above)
Comment (spireken) [reply Applications/Spire/Include/Spire/Ui/DecimalInputWidget.hpp:23]: I might be able to get rid of the signals, but there's a few changes related to editing that go beyond just a style change.

Comment (kranar) [reply Applications/Spire/Include/Spire/Ui/DecimalInputWidget.hpp:23]: Okay, keep all the logic then, remove the signals.

File: Applications/Spire/Include/Spire/Ui/QuantityInputWidget.hpp (hunk 10 above)
Comment (spireken) [reply Applications/Spire/Include/Spire/Ui/QuantityInputWidget.hpp:9]: I did, but it involves defining the valueFromText, textFromValue, and validate functions because the default implementations do nothing, among other tasks to get it to work. I thought a double could represent the types we need so modifying the QDoubleSpinBox would be the most straightforward way to implement this.

Comment (kranar) [reply Applications/Spire/Include/Spire/Ui/QuantityInputWidget.hpp:9]: Okay sounds good.

File: Applications/Spire/Source/Ui/QuantityInputWidget.cpp (hunk 12 above)
Comment (spireken) [reply Applications/Spire/Source/Ui/QuantityInputWidget.cpp:37]: I borrowed this from the CustomVariantItemDelegate::displayText(...) function that converts a Quantity to a string. Would I have to go Quantity -> QString -> double to convert this properly?

Also, how is 1.1 not a valid double? Or is it just the context?

Comment (kranar) [reply Applications/Spire/Source/Ui/QuantityInputWidget.cpp:37]: There is no way to represent the value ```1.1``` as a double.

I can look into this a bit later today but I was hoping you could set the value of the ```QDoubleSpinBox``` using a ```QString``` instead of a ```double```.

Comment (spireken) [reply Applications/Spire/Source/Ui/QuantityInputWidget.cpp:37]: I can create a getter to access the underlying QLineEdit to set the current text, but there's still the issue of setting the range. There is a private implementation (QAbstractSpinBoxPrivate) that has a setRange() function that takes QVariants.

Comment (kranar) [reply Applications/Spire/Source/Ui/QuantityInputWidget.cpp:37]: Setting the range isn't a concern since there's no use case where a range is set to more than 15 decimal places.

Comment (spireken) [reply Applications/Spire/Source/Ui/QuantityInputWidget.cpp:37]: Sorry, the range for the spin box means the acceptable input range from the minimum to maximum value.

Would it be possible to just use the conversion functions given by the QDoubleSpinBox? Converting the Quantity to a QString then passing it to QDoubleSpinBox::valueFromText() would give the double that the input would return from the given text.

Comment (kranar) [reply Applications/Spire/Source/Ui/QuantityInputWidget.cpp:37]: Yep I got what you meant by range, it's not an issue for our use case.

Your solution of using ```valueFromText``` sounds good. Go with it.

**Comment by spireken** [comment]
The QDoubleSpinBox::valueFromText() function bounds the value to the current range, which is why I 'faked' an unlimited range in the helper function.
